File: clojure/cljfind/src/cljfind/filetypes.clj
;;; ############################################################################
;;;
;;; filetypes.clj
;;;
;;; Utility functions for getting file extension and determining file type
;;;
;;; ############################################################################

(ns cljfind.filetypes
  #^{:author "Cary Clark",
     :doc "Module to provide file-related utility functions"}
  (:import (java.io File))
  (:require [clojure.java.io :as io])
  (:require [clojure.java.jdbc :as jdbc])
  (:require [clojure.data.json :as json])
  (:use [clojure.set :only (union)]
        [clojure.string :only (lower-case)]
        [cljfind.config :only (XFINDDB)]
        [cljfind.fileutil :only (get-ext get-name)]))

(def ^:const ^String ARCHIVE "archive")
(def ^:const ^String AUDIO "audio")
(def ^:const ^String BINARY "binary")
(def ^:const ^String CODE "code")
(def ^:const ^String FONT "font")
(def ^:const ^String IMAGE "image")
(def ^:const ^String TEXT "text")
(def ^:const ^String VIDEO "video")
(def ^:const ^String XML "xml")

(def ^:const FILE-TYPES
  [:unknown :archive :audio :binary :code :font :image :text :video :xml])

(def ^:const db
  {:classname   "org.sqlite.JDBC"
   :subprotocol "sqlite"
   :subname     XFINDDB})

(def ^:const ^String FILE-NAME-QUERY "SELECT file_type_id FROM file_name WHERE name = ?")
(def ^:const ^String FILE-EXT-QUERY "SELECT file_type_id FROM file_extension WHERE extension = ?")

(defn get-file-type-for-file-name [^String file-name]
  (let [file-type-id (jdbc/query db [FILE-NAME-QUERY file-name] {:row-fn :file_type_id})]
    (if (or (nil? file-type-id) (empty? file-type-id))
      :unknown
      (nth FILE-TYPES (- (first file-type-id) 1)))))

(defn get-file-type-for-extension [^String file-ext]
  (let [file-type-id (jdbc/query db [FILE-EXT-QUERY file-ext] {:row-fn :file_type_id})]
    (if (or (nil? file-type-id) (empty? file-type-id))
      :unknown
      (nth FILE-TYPES (- (first file-type-id) 1)))))

(defn get-file-type [f]
  (let [file-name (get-name f)
        file-type-for-file-name (get-file-type-for-file-name file-name)]
    (if (not= :unknown file-type-for-file-name)
      file-type-for-file-name
      (get-file-type-for-extension (get-ext f)))))

(defn archive-file? [f]
  (= (get-file-type f) :archive))

(defn audio-file? [f]
  (= (get-file-type f) :audio))

(defn binary-file? [f]
  (= (get-file-type f) :binary))

(defn code-file? [f]
  (= (get-file-type f) :code))

(defn font-file? [f]
  (= (get-file-type f) :font))

(defn image-file? [f]
  (= (get-file-type f) :image))

(defn text-file? [f]
  (let [file-type (get-file-type f)]
    (or
     (= file-type :text)
     (= file-type :code)
     (= file-type :xml))))

(defn video-file? [f]
  (= (get-file-type f) :video))

(defn xml-file? [f]
  (= (get-file-type f) :xml))

(defn unknown-file? [f]
  (= :unknown (get-file-type f)))

(defn from-name [^String name]
  (let [lname (lower-case name)]
    (cond
      (= ARCHIVE lname) :archive
      (= AUDIO lname) :audio
      (= BINARY lname) :binary
      (= CODE lname) :code
      (= FONT lname) :font
      (= IMAGE lname) :image
      (= TEXT lname) :text
      (= VIDEO lname) :video
      (= XML lname) :xml
      :else :unknown)))

(defn to-name ^String [ft]
  (cond
    (= :archive ft) ARCHIVE
    (= :audio ft) AUDIO
    (= :binary ft) BINARY
    (= :code ft) CODE
    (= :font ft) FONT
    (= :image ft) IMAGE
    (= :text ft) TEXT
    (= :video ft) VIDEO
    (= :xml ft) XML
    :else "unknown"))

File: haskell/hsfind/src/HsFind/FileTypes.hs
{-# LANGUAGE NoMonomorphismRestriction, OverloadedStrings #-}
module HsFind.FileTypes
  ( FileType(..)
  , getFileTypeForName
  , getFileTypeName
  , getDbFileType
  , getDbFileTypes
  , isArchiveFilePath
  , isAudioFilePath
  , isBinaryFilePath
  , isCodeFilePath
  , isFontFilePath
  , isImageFilePath
  , isTextFilePath
  , isVideoFilePath
  , isXmlFilePath
  , isUnknownFilePath
  ) where

import Data.Char (toLower)
import Data.List (intercalate, nub, partition)
import qualified Data.Text as T
import System.FilePath (takeFileName)

import Database.SQLite.Simple

import HsFind.FileUtil (getExtension)
import Data.Maybe (mapMaybe)

data FileType = Unknown
              | Archive
              | Audio
              | Binary
              | Code
              | Font
              | Image
              | Text
              | Video
              | Xml
  deriving (Show, Eq, Ord)

fileTypes :: [FileType]
fileTypes = [Unknown, Archive, Audio, Binary, Code, Font, Image, Text, Video, Xml]

getFileTypeForName :: String -> FileType
getFileTypeForName typeName =
  case lower typeName of
    "archive" -> Archive
    "audio" -> Audio
    "binary" -> Binary
    "code" -> Code
    "font" -> Font
    "image" -> Image
    "text" -> Text
    "video" -> Video
    "xml" -> Xml
    _ -> Unknown
  where lower = map toLower

getFileTypeName :: FileType -> String
getFileTypeName ft =
  case ft of
    Archive -> "archive"
    Audio -> "audio"
    Binary -> "binary"
    Code -> "code"
    Font -> "font"
    Image -> "image"
    Text -> "text"
    Video -> "video"
    Xml -> "xml"
    _ -> "unknown"

batchSize :: Int
batchSize = 500

getDbFileTypesForQueryAndElems :: Connection -> [String] -> [String] -> IO [(String, FileType)]
getDbFileTypesForQueryAndElems conn queryElems [] = return []
getDbFileTypesForQueryAndElems conn queryElems elems = do
  let uniqElems = nub elems
  getFileTypesForElems uniqElems []
  where getFileTypesForElems :: [String] -> [(String, FileType)] -> IO [(String, FileType)]
        getFileTypesForElems elems elemsAndTypes =
          case elems of
            [] -> return elemsAndTypes
            _ -> do
              let elemBatch = take batchSize elems
              let placeholders = intercalate ", " $ replicate (length elemBatch) "?" :: String
              let selectStr = concatMap (\q -> if q == "?" then placeholders else q) queryElems :: String
              let selectText = T.pack selectStr
              let selectQuery = Query {fromQuery=selectText}
              rows <- query conn selectQuery elemBatch :: IO [(String, Int)]
              let dbElemBatchTypes = map (\(e, i) -> (e, fileTypes !! (i - 1))) rows
              let elemBatchAndTypes = getDbFileTypesForElems elemBatch dbElemBatchTypes []
              getFileTypesForElems (drop batchSize elems) $ elemsAndTypes ++ elemBatchAndTypes
        getDbFileTypesForElems :: [String] -> [(String, FileType)] -> [(String, FileType)] -> [(String, FileType)]
        getDbFileTypesForElems elems dbFileElemTypes elemsAndTypes =
          case elems of
            [] -> elemsAndTypes
            (e:elems') ->
              case lookup e dbFileElemTypes of
                Just ft -> getDbFileTypesForElems elems' dbFileElemTypes $ elemsAndTypes ++ [(e, ft)]
                Nothing -> getDbFileTypesForElems elems' dbFileElemTypes $ elemsAndTypes ++ [(e, Unknown)]

getDbFileTypesForFileNames :: Connection -> [FilePath] -> IO [(String, FileType)]
getDbFileTypesForFileNames conn [] = return []
getDbFileTypesForFileNames conn filePaths = do
  let queryElems = ["SELECT name, file_type_id FROM file_name WHERE name IN (", "?", ")"]
  let fileNames = map takeFileName filePaths
  getDbFileTypesForQueryAndElems conn queryElems fileNames

getDbFileTypeForFileName :: Connection -> FilePath -> IO FileType
getDbFileTypeForFileName conn fileName = do
  row <- query conn "SELECT file_type_id FROM file_name WHERE name = ?" (Only (fileName :: String)) :: IO [Only Int]
  case row of
    [Only i] -> return $ fileTypes !! (i - 1)
    _ -> return Unknown

getDbFileTypesForExtensions :: Connection -> [String] -> IO [(String, FileType)]
getDbFileTypesForExtensions conn [] = return []
getDbFileTypesForExtensions conn fileExts = do
  let queryElems = ["SELECT extension, file_type_id FROM file_extension WHERE extension IN (", "?", ")"]
  getDbFileTypesForQueryAndElems conn queryElems fileExts

getDbFileTypeForExtension :: Connection -> String -> IO FileType
getDbFileTypeForExtension conn fileExt = do
  row <- query conn "SELECT file_type_id FROM file_extension WHERE extension = ?" (Only (fileExt :: String)) :: IO [Only Int]
  case row of
    [Only i] -> return $ fileTypes !! (i - 1)
    _ -> return Unknown

getDbFileType :: Connection -> FilePath -> IO FileType
getDbFileType conn f = do
  ft <- getDbFileTypeForFileName conn (takeFileName f)
  case ft of
    Unknown -> case getExtension f of
                  Just x -> getDbFileTypeForExtension conn x
                  Nothing -> return Unknown
    _  -> return ft

-- getDbFileTypes :: Connection -> [FilePath] -> IO [FileType]
-- getDbFileTypes conn filePaths = do
--   mapM (getDbFileType conn) filePaths

getDbFileTypes :: Connection -> [FilePath] -> IO [FileType]
getDbFileTypes conn filePaths = do
  dbFileTypesForFileNames <- getDbFileTypesForFileNames conn $ map takeFileName filePaths
  let (knownTypesForFileNames, unknownTypesForFileNames) = partition (\(name, ft) -> ft /= Unknown) dbFileTypesForFileNames
  let unknownFileNames = map fst unknownTypesForFileNames
  dbFileTypesForExts <- getDbFileTypesForExtensions conn $ mapMaybe getExtension unknownFileNames
  return $ getFilePathsWithFileTypes filePaths knownTypesForFileNames dbFileTypesForExts []
  where getFilePathsWithFileTypes :: [FilePath] -> [(FilePath, FileType)] -> [(FilePath, FileType)] -> [(FilePath, FileType)] -> [FileType]
        getFilePathsWithFileTypes paths knownTypesForFileNames dbFileTypesForExts pathsAndTypes =
          case paths of
            [] -> map snd pathsAndTypes
            (path:paths') ->
              case lookup (takeFileName path) knownTypesForFileNames of
                Just ft -> getFilePathsWithFileTypes paths' knownTypesForFileNames dbFileTypesForExts $ pathsAndTypes ++ [(path, ft)]
                Nothing ->
                  case getExtension path of
                    Just ext ->
                      case lookup ext dbFileTypesForExts of
                        Just ft -> getFilePathsWithFileTypes paths' knownTypesForFileNames dbFileTypesForExts $ pathsAndTypes ++ [(path, ft)]
                        Nothing -> getFilePathsWithFileTypes paths' knownTypesForFileNames dbFileTypesForExts $ pathsAndTypes ++ [(path, Unknown)]
                    Nothing -> getFilePathsWithFileTypes paths' knownTypesForFileNames dbFileTypesForExts $ pathsAndTypes ++ [(path, Unknown)]

hasFilePathFileType :: Connection -> FilePath -> FileType -> IO Bool
hasFilePathFileType conn fp ft = do
  dbFileType <- getDbFileType conn fp
  return $ dbFileType == ft

isUnknownFilePath :: Connection -> FilePath -> IO Bool
isUnknownFilePath conn fp = do
  hasFilePathFileType conn fp Unknown

isArchiveFilePath :: Connection -> FilePath -> IO Bool
isArchiveFilePath conn fp = do
  hasFilePathFileType conn fp Archive

isAudioFilePath :: Connection -> FilePath -> IO Bool
isAudioFilePath conn fp = do
  hasFilePathFileType conn fp Audio

isBinaryFilePath :: Connection -> FilePath -> IO Bool
isBinaryFilePath conn fp = do
  hasFilePathFileType conn fp Binary

isCodeFilePath :: Connection -> FilePath -> IO Bool
isCodeFilePath conn fp = do
  hasFilePathFileType conn fp Code

isFontFilePath :: Connection -> FilePath -> IO Bool
isFontFilePath conn fp = do
  hasFilePathFileType conn fp Font

isImageFilePath :: Connection -> FilePath -> IO Bool
isImageFilePath conn fp = do
  hasFilePathFileType conn fp Image

isTextFilePath :: Connection -> FilePath -> IO Bool
isTextFilePath conn fp = do
  dbFileType <- getDbFileType conn fp
  return $ dbFileType `elem` [Text, Code, Xml]

isVideoFilePath :: Connection -> FilePath -> IO Bool
isVideoFilePath conn fp = do
  hasFilePathFileType conn fp Video

isXmlFilePath :: Connection -> FilePath -> IO Bool
isXmlFilePath conn fp = do
  hasFilePathFileType conn fp Xml
